tenant/repository_test.go

```diff
@@ -143,6 +146,79 @@ func (s *TenantServiceTestSuite) TestLookupTenantByNamespace() {
 
 }
 
+func (s *TenantServiceTestSuite) TestGetAllTenantsToUpdate() {
+	s.T().Run("returns all tenants", func(t *testing.T) {
+		// given
+		controller.Commit = "123abc"
+		testdoubles.SetTemplateVersions()
+		tf.FillDB(t, s.DB, 3, false, "ready", environment.DefaultEnvTypes...)
+		svc := tenant.NewDBService(s.DB)
+
+		// when
+		result, err := svc.GetTenantsToUpdate(testdoubles.GetMappedVersions(environment.DefaultEnvTypes...), 10, "xyz")
+
+		// then
+		assert.NoError(t, err)
+		assert.Len(t, result, 3)
+	})
+
+	s.T().Run("returns only the limited number of tenants", func(t *testing.T) {
+		// given
+		controller.Commit = "123abc"
+		testdoubles.SetTemplateVersions()
+		tf.FillDB(t, s.DB, 10, false, "ready", environment.DefaultEnvTypes...)
+		svc := tenant.NewDBService(s.DB)
+
+		// when
+		result, err := svc.GetTenantsToUpdate(testdoubles.GetMappedVersions(environment.DefaultEnvTypes...), 5, "xyz")
+
+		// then
+		assert.NoError(t, err)
+		assert.Len(t, result, 5)
+	})
+}
+
+func (s *TenantServiceTestSuite) TestGetSubsetOfFailedTenantsToUpdate() {
+	s.T().Run("returns only those tenants whose namespaces have different updated_by", func(t *testing.T) {
+		// given
+		testdoubles.SetTemplateVersions()
+		controller.Commit = "123abc"
+		previouslyFailed := tf.FillDB(t, s.DB, 1, false, "failed", environment.DefaultEnvTypes...)
+		controller.Commit = "234bcd"
+		tf.FillDB(t, s.DB, 6, false, "failed", environment.DefaultEnvTypes...)
+
+		svc := tenant.NewDBService(s.DB)
+
+		// when
+		result, err := svc.GetTenantsToUpdate(testdoubles.GetMappedVersions(environment.DefaultEnvTypes...), 10, "234bcd")
+
+		// then
+		assert.NoError(t, err)
+		assert.Len(t, result, 1)
+		assert.Equal(t, previouslyFailed.Tenants[0].ID, result[0].ID)
+	})
+}
+
+func (s *TenantServiceTestSuite) TestGetSubsetOfTenantsThatAreOutdatedToUpdate() {
+	s.T().Run("returns only those tenants whose namespaces have different version", func(t *testing.T) {
+		// given
+		testdoubles.SetTemplateVersions()
+		controller.Commit = "123abc"
+		outdated := tf.FillDB(t, s.DB, 1, false, "ready", environment.DefaultEnvTypes...)
+		tf.FillDB(t, s.DB, 6, true, "ready", environment.DefaultEnvTypes...)
+
+		svc := tenant.NewDBService(s.DB)
+
+		// when
+		result, err := svc.GetTenantsToUpdate(testdoubles.GetMappedVersions(environment.DefaultEnvTypes...), 10, "234bcd")
+
+		// then
+		assert.NoError(t, err)
+		assert.Len(t, result, 1)
+		assert.Equal(t, outdated.Tenants[0].ID, result[0].ID)
+	})
+}
+
```

Comment:
Let's also add a test where we have let's say 11 tenants to update. Then GetTenantsToUpdate for the first 5 tenants and check that they are expected ones. Then mark them as "updated" in DB and get the next 5. Check they are expected then "update" and get the next 5 - should get the last one remaining. So, you will imitate the whole cycle.

Reply:
added test here 0d2b231  I covered both cases there - when tenants are successfully updated and when the update failed